README.md
# [**@koa/send**](https://github.com/koajs/send)


[![NPM version][npm-image]][npm-url]
![Build Status][github-action-image]
[![Test coverage][coveralls-image]][coveralls-url]
[![License][license-image]][license-url]
[![Downloads][downloads-image]][downloads-url]

[npm-image]: https://img.shields.io/npm/v/@koa/send.svg?style=flat-square
[npm-url]: https://npmjs.org/package/@koa/send
[github-action-image]: https://github.com/koajs/send/actions/workflows/ci.yml/badge.svg?style=flat-square
[coveralls-image]: https://img.shields.io/coveralls/koajs/send.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/koajs/send?branch=master
[license-image]: http://img.shields.io/npm/l/koa-send.svg?style=flat-square
[license-url]: LICENSE
[downloads-image]: http://img.shields.io/npm/dm/koa-send.svg?style=flat-square
[downloads-url]: https://npmjs.org/package/koa-send

Koa static file serving middleware.

## Install

[![NPM](https://nodei.co/npm/@koa/send.png?downloads=true)](https://nodei.co/npm/@koa/send)

```js
$ npm i @koa/send
```

## Options

- `maxage` Browser cache max-age in milliseconds. (defaults to `0`).
- `immutable` Tell the browser the resource is immutable and can be cached indefinitely. (defaults to `false`).
- `hidden` Allow transfer of hidden files. (defaults to `false`).
- [`root`](#root-path) Root directory to restrict file access.
- `index` Name of the index file to serve automatically when visiting the root location. (defaults to none).
- `gzip` Try to serve the gzipped version of a file automatically when `gzip` is supported by a client and if the requested file with `.gz` extension exists. (defaults to `true`).
- `brotli` Try to serve the brotli version of a file automatically when `brotli` is supported by a client and if the requested file with `.br` extension exists. (defaults to `true`).
- `format` If not `false` (defaults to `true`), format the path to serve static file servers and not require a trailing slash for directories, so that you can do both `/directory` and `/directory/`.
- [`setHeaders`](#setheaders) Function to set custom headers on response.
- `extensions` Try to match extensions from passed array to search for file when no extension is sufficed in URL. First found is served. (defaults to `false`)

### Root path

Note that `root` is required, defaults to `''` and will be resolved,
removing the leading `/` to make the path relative and this
path must not contain "..", protecting developers from
concatenating user input. If you plan on serving files based on
user input supply a `root` directory from which to serve from.

For example to serve files from `./public`:

```js
app.use(async (ctx) => {
  await send(ctx, ctx.path, { root: __dirname + "/public" });
});
```

To serve developer specified files:

```js
app.use(async (ctx) => {
  await send(ctx, "path/to/my.js");
});
```

### setHeaders

The function is called as `fn(res, path, stats)`, where the arguments are:

- `res`: the response object.
- `path`: the resolved file path that is being sent.
- `stats`: the stats object of the file that is being sent.

You should only use the `setHeaders` option when you wish to edit the `Cache-Control` or `Last-Modified` headers, because doing it before is useless (it's overwritten by `send`), and doing it after is too late because the headers are already sent.

If you want to edit any other header, simply set them before calling `send`.

## Example

```js
const send = require("@koa/send");
const Koa = require("koa");
const app = new Koa();

// $ GET /package.json
// $ GET /

app.use(async (ctx) => {
  if ("/" == ctx.path) return (ctx.body = "Try GET /package.json");
  await send(ctx, ctx.path);
});

app.listen(3000);
console.log("listening on port 3000");
```

## License

[MIT](/LICENSE)


History.md
5.0.1 / 2020-07-08
==================

 * Replace mz with custom promisified wrappers (#140)
 * fix(jsdoc): send return type (#141)
 * fix extensions options on windows machines (#79) (#122)
 * Revert "Using RegExp replace loop in isPathHidden (#109)" (#138)
 * Using RegExp replace loop in isPathHidden (#109)
 * Clean up / Pipeline update ... (#135)
 * Update .travis.yml
 * Fix Tests & coverage (#132)
 * Fix license (#134)
 * travis: add node@11

5.0.0 / 2018-06-19
==================

 * bump deps
 * fix bug that set Content-Type not working. (#105)

4.1.2 / 2017-12-14
==================

 * Fix issue with dots in path when using extensions array (#92)

4.1.1 / 2017-09-25
==================

 * Fix brotli support, closes #83
 * Fix tests

4.1.0 / 2017-04-26
==================

 * Add support for Cache-Control: immutable with a new "immutable" option
 * Added serving of brotli versions of files

4.0.0 / 2017-04-09
==================

 * throw error if file not exists, closes #43
 * remove co, use async function
 * bump deps

3.3.0 / 2017-01-10
==================

 * add extensions option
 * bump deps

3.2.0 / 2016-03-23
==================

 * add setHeaders option

3.1.1 / 2016-03-04
==================

 * bump deps
 * simplify scripts

3.1.0 / 2015-10-24
==================

 * return a promise instead of a generator
 * fix: split path by path.sep instead of slash
 * fix: strip root correctly on windows
 * tests: resolve paths for windows

3.0.1 / 2015-10-23
==================

 * fix stats info when path does not finish with slash and format is enabled, closes #34

3.0.0 / 2015-10-21
==================

 * bump deps
 * format option defaults to true
 * fix: enable format only index exists
 * simplify the declarations of `format`, `gzip`

2.0.1 / 2015-10-14
==================

 * fix judgement of trailing slash

2.0.0 / 2015-10-14
==================

 * serve directories without a trailing slash, closes #27
 * when .hidden option is set, also check hidden directories, closes #17
 * bump deps: mz@2, mocha@2, koa@1
 * use resolve-path, closes #9
 * gzip version of file should not be automatically sent
 * fix test: gzip.json.gz is not valid json data

1.3.1 / 2014-09-08
==================

 * add .maxAge alias

1.3.0 / 2014-09-07
==================

 * add automatically check and serve `.gz` files
 * remove `finished` dependency
 * refactor with `mz`

1.2.3 / 2014-02-11
==================

 * fix malicious path in windows
 * update finished
 * make assert message better

1.2.2 / 2014-01-07
==================

 * fix: ignore directories instead of crashing koa

1.2.1 / 2014-01-02
==================

 * add `content-length` header

1.2.0 / 2013-12-27
==================

 * add `maxage` option

1.1.2 / 2013-12-22
==================

 * replace deprecated ctx.error() with ctx.throw()

1.1.1 / 2013-12-20
==================

 * use: on-socket-error

1.1.0 / 2013-12-19
==================

 * add: `send` now returns the file path if sent
 * add: destroy streams on socket errors to prevent fd leaks


test/index.js

const request = require('supertest')
const send = require('..')
const path = require('path')
const Koa = require('koa')
const assert = require('assert')
const decompress = require('iltorb').decompress

describe('send(ctx, file)', function () {
  describe('with no .root', function () {
    describe('when the path is absolute', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, path.join(__dirname, '/fixtures/hello.txt'))
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })

      it('should throw 404 error', function (done) {
        const app = new Koa()

        let error
        app.use(async (ctx) => {
          try {
            await send(ctx, path.join(__dirname, '/fixtures/hello.txt'))
          } catch (err) {
            error = err
          }
        })

        request(app.listen())
          .get('/')
          .expect(404, () => {
            assert.equal(error.status, 404)
            done()
          })
      })
    })

    describe('when the path is relative', function () {
      it('should 200', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello.txt')
        })

        request(app.listen())
          .get('/')
          .expect(200)
          .expect('world', done)
      })
    })

    describe('when the path contains ..', function () {
      it('should 403', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/../fixtures/hello.txt')
        })

        request(app.listen())
          .get('/')
          .expect(403, done)
      })
    })
  })

  describe('with .root', function () {
    describe('when the path is absolute', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test/fixtures' }
          await send(ctx, path.join(__dirname, '/fixtures/hello.txt'), opts)
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when the path is relative and exists', function () {
      it('should serve the file', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test/fixtures' }
          await send(ctx, 'hello.txt', opts)
        })

        request(app.listen())
          .get('/')
          .expect(200)
          .expect('world', done)
      })
    })

    describe('when the path is relative and does not exist', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test/fixtures' }
          await send(ctx, 'something', opts)
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when the path resolves above the root', function () {
      it('should 403', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test/fixtures' }
          await send(ctx, '../../package.json', opts)
        })

        request(app.listen())
          .get('/')
          .expect(403, done)
      })
    })

    describe('when the path resolves within root', function () {
      it('should 403', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test/fixtures' }
          await send(ctx, '../../test/fixtures/world/index.html', opts)
        })

        request(app.listen())
          .get('/')
          .expect(403, done)
      })
    })
  })

  describe('with .index', function () {
    describe('when the index file is present', function () {
      it('should serve it', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test', index: 'index.html' }
          await send(ctx, 'fixtures/world/', opts)
        })

        request(app.listen())
          .get('/')
          .expect(200)
          .expect('html index', done)
      })

      it('should serve it', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const opts = { root: 'test/fixtures/world', index: 'index.html' }
          await send(ctx, ctx.path, opts)
        })

        request(app.listen())
          .get('/')
          .expect(200)
          .expect('html index', done)
      })
    })
  })

  describe('when path is not a file', function () {
    it('should 404', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        await send(ctx, '/test')
      })

      request(app.listen())
        .get('/')
        .expect(404, done)
    })

    it('should return undefined if format is set to false', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        const sent = await send(ctx, '/test', { format: false })
        assert.equal(sent, undefined)
      })

      request(app.listen())
        .get('/')
        .expect(404, done)
    })
  })

  describe('when path is a directory', function () {
    it('should 404', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        await send(ctx, '/test/fixtures')
      })

      request(app.listen())
        .get('/')
        .expect(404, done)
    })
  })

  describe('when path does not finish with slash and format is disabled', function () {
    it('should 404', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        const opts = { root: 'test', index: 'index.html', format: false }
        await send(ctx, 'fixtures/world', opts)
      })

      request(app.listen())
        .get('/world')
        .expect(404, done)
    })

    it('should 404', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        const opts = { root: 'test', index: 'index.html', format: false }
        await send(ctx, 'fixtures/world', opts)
      })

      request(app.listen())
        .get('/world')
        .expect(404, done)
    })
  })

  describe('when path does not finish with slash and format is enabled', function () {
    it('should 200', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        const opts = { root: 'test', index: 'index.html' }
        await send(ctx, 'fixtures/world', opts)
      })

      request(app.listen())
        .get('/')
        .expect('content-type', 'text/html; charset=utf-8')
        .expect('content-length', '10')
        .expect(200, done)
    })

    it('should 404 if no index', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        const opts = { root: 'test' }
        await send(ctx, 'fixtures/world', opts)
      })

      request(app.listen())
        .get('/')
        .expect(404, done)
    })
  })

  describe('when path is malformed', function () {
    it('should 400', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        await send(ctx, '/%')
      })

      request(app.listen())
        .get('/')
        .expect(400, done)
    })

    it('should display which path causes decode error', function (done) {
      const app = new Koa()
      const path = '%public_folder%/favicon.ico'

      app.use(async (ctx) => {
        await send(ctx, '/' + path)
      })

      request(app.listen())
        .get('/')
        .expect('failed to decode ' + path, done)
    })
  })

  describe('when path is a file', function () {
    it('should return the path', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        const p = '/test/fixtures/user.json'
        const sent = await send(ctx, p)
        assert.equal(sent, path.join(__dirname, '/fixtures/user.json'))
      })

      request(app.listen())
        .get('/')
        .expect(200, done)
    })

    describe('or .gz version when requested and if possible', function () {
      it('should return path', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json')
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'deflate, identity')
          .expect('Content-Length', '18')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })

      it('should return .gz path (gzip option defaults to true)', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json')
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'gzip, deflate, identity')
          .expect('Content-Length', '48')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })

      it('should return .gz path when gzip option is turned on', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json', { gzip: true })
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'gzip, deflate, identity')
          .expect('Content-Length', '48')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })

      it('should not return .gz path when gzip option is false', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json', { gzip: false })
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'gzip, deflate, identity')
          .expect('Content-Length', '18')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })
    })

    describe('or .br version when requested and if possible', function () {
      function parser (res, cb) {
        const chunks = []
        res.on('data', chunk => {
          chunks.push(chunk)
        })
        res.on('end', () => {
          decompress(Buffer.concat(chunks), (err, data) => {
            cb(err, data.toString())
          })
        })
      }

      it('should return path', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json')
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'deflate, identity')
          .expect('Content-Length', '18')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })

      it('should return .br path (brotli option defaults to true)', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json')
        })

        request(app.listen())
          .get('/')
          .parse(parser)
          .set('Accept-Encoding', 'br, deflate, identity')
          .expect('Content-Length', '22')
          .expect(200)
          .then(({body}) => {
            assert.deepStrictEqual(body, '{ "name": "tobi" }')
            done()
          })
      })

      it('should return .br path when brotli option is turned on', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json', { brotli: true })
        })

        request(app.listen())
          .get('/')
          .parse(parser)
          .set('Accept-Encoding', 'br, deflate, identity')
          .expect('Content-Length', '22')
          .expect(200)
          .then(({body}) => {
            assert.deepStrictEqual(body, '{ "name": "tobi" }')
            done()
          })
      })

      it('should not return .br path when brotli option is false', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json', { brotli: false })
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'br, deflate, identity')
          .expect('Content-Length', '18')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })

      it('should return .gz path when brotli option is turned off', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, '/test/fixtures/gzip.json', { brotli: false })
        })

        request(app.listen())
          .get('/')
          .set('Accept-Encoding', 'br, gzip, deflate, identity')
          .expect('Content-Length', '48')
          .expect('{ "name": "tobi" }')
          .expect(200, done)
      })
    })

    describe('and max age is specified', function () {
      it('should set max-age in seconds', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const p = '/test/fixtures/user.json'
          const sent = await send(ctx, p, { maxage: 5000 })
          assert.equal(sent, path.join(__dirname, '/fixtures/user.json'))
        })

        request(app.listen())
          .get('/')
          .expect('Cache-Control', 'max-age=5')
          .expect(200, done)
      })

      it('should truncate fractional values for max-age', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const p = '/test/fixtures/user.json'
          const sent = await send(ctx, p, { maxage: 1234 })
          assert.equal(sent, path.join(__dirname, '/fixtures/user.json'))
        })

        request(app.listen())
          .get('/')
          .expect('Cache-Control', 'max-age=1')
          .expect(200, done)
      })
    })

    describe('and immutable is specified', function () {
      it('should set the immutable directive', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          const p = '/test/fixtures/user.json'
          const sent = await send(ctx, p, { immutable: true, maxage: 31536000000 })
          assert.equal(sent, path.join(__dirname, '/fixtures/user.json'))
        })

        request(app.listen())
          .get('/')
          .expect('Cache-Control', 'max-age=31536000,immutable')
          .expect(200, done)
      })
    })
  })

  describe('.immutable option', function () {
    describe('when trying to get a non-existent file', function () {
      it('should not set the Cache-Control header', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/does-not-exist.json', { immutable: true })
        })

        request(app.listen())
          .get('/')
          .expect((res) => {
            assert.equal(res.header['cache-control'], undefined)
          })
          .expect(404, done)
      })
    })
  })

  describe('.hidden option', function () {
    describe('when trying to get a hidden file', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/.hidden')
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when trying to get a file from a hidden directory', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/.private/id_rsa.txt')
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when trying to get a hidden file and .hidden check is turned off', function () {
      it('should 200', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/.hidden', { hidden: true })
        })

        request(app.listen())
          .get('/')
          .expect(200, done)
      })
    })
  })

  describe('.extensions option', function () {
    describe('when trying to get a file without extension with no .extensions sufficed', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello')
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when trying to get a file without extension with no matching .extensions', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello', { extensions: ['json', 'htm', 'html'] })
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when trying to get a file without extension with non array .extensions', function () {
      it('should 404', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello', { extensions: {} })
        })

        request(app.listen())
          .get('/')
          .expect(404, done)
      })
    })

    describe('when trying to get a file without extension with non string array .extensions', function () {
      it('throws if extensions is not array of strings', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello', { extensions: [2, {}, []] })
        })

        request(app.listen())
          .get('/')
          .expect(500)
          .end(done)
      })
    })

    describe('when trying to get a file without extension with matching .extensions sufficed first matched should be sent', function () {
      it('should 200 and application/json', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/user', { extensions: ['html', 'json', 'txt'] })
        })

        request(app.listen())
          .get('/')
          .expect(200)
          .expect('Content-Type', /application\/json/)
          .end(done)
      })
    })

    describe('when trying to get a file without extension with matching .extensions sufficed', function () {
      it('should 200', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello', { extensions: ['txt'] })
        })

        request(app.listen())
          .get('/')
          .expect(200, done)
      })
    })

    describe('when trying to get a file without extension with matching doted .extensions sufficed', function () {
      it('should 200', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/hello', { extensions: ['.txt'] })
        })

        request(app.listen())
          .get('/')
          .expect(200, done)
      })
    })

    describe('when trying to get a file without extension with matching .extensions sufficed with other dots in path', function () {
      it('should 200', function (done) {
        const app = new Koa()

        app.use(async (ctx) => {
          await send(ctx, 'test/fixtures/some.path/index', { extensions: ['json'] })
        })

        request(app.listen())
          .get('/')
          .expect(200, done)
      })
    })
  })

  it('should set the Content-Type', function (done) {
    const app = new Koa()

    app.use(async (ctx) => {
      await send(ctx, '/test/fixtures/user.json')
    })

    request(app.listen())
      .get('/')
      .expect('Content-Type', /application\/json/)
      .end(done)
  })

  it('should set the Content-Length', function (done) {
    const app = new Koa()

    app.use(async (ctx) => {
      await send(ctx, '/test/fixtures/user.json')
    })

    request(app.listen())
      .get('/')
      .expect('Content-Length', '18')
      .end(done)
  })

  it('should set the Content-Type', function (done) {
    const app = new Koa()

    const testFilePath = path.normalize('/test/fixtures/world/index.html')

    app.use(async (ctx) => {
      ctx.type = 'text/plain'
      await send(ctx, testFilePath)
    })

    request(app.listen())
      .get('/')
      .expect('Content-Type', /text\/plain/)
      .end(done)
  })

  it('should set Last-Modified', function (done) {
    const app = new Koa()

    app.use(async (ctx) => {
      await send(ctx, '/test/fixtures/user.json')
    })

    request(app.listen())
      .get('/')
      .expect('Last-Modified', /GMT/)
      .end(done)
  })

  describe('with setHeaders', function () {
    it('throws if setHeaders is not a function', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        await send(ctx, '/test/fixtures/user.json', {
          setHeaders: 'foo'
        })
      })

      request(app.listen())
        .get('/')
        .expect(500)
        .end(done)
    })

    it('should not edit already set headers', function (done) {
      const app = new Koa()

      const testFilePath = '/test/fixtures/user.json'
      const normalizedTestFilePath = path.normalize(testFilePath)

      app.use(async (ctx) => {
        await send(ctx, testFilePath, {
          setHeaders: function (res, path, stats) {
            assert.equal(path.substr(-normalizedTestFilePath.length), normalizedTestFilePath)
            assert.equal(stats.size, 18)
            assert(res)

            // these can be set
            res.setHeader('Cache-Control', 'max-age=0,must-revalidate')
            res.setHeader('Last-Modified', 'foo')
            // this one can not
            res.setHeader('Content-Length', 9000)
          }
        })
      })

      request(app.listen())
        .get('/')
        .expect(200)
        .expect('Cache-Control', 'max-age=0,must-revalidate')
        .expect('Last-Modified', 'foo')
        .expect('Content-Length', '18')
        .end(done)
    })

    it('should correctly pass through regarding usual headers', function (done) {
      const app = new Koa()

      app.use(async (ctx) => {
        await send(ctx, '/test/fixtures/user.json', {
          setHeaders: () => {}
        })
      })

      request(app.listen())
        .get('/')
        .expect(200)
        .expect('Cache-Control', 'max-age=0')
        .expect('Content-Length', '18')
        .expect('Last-Modified', /GMT/)
        .end(done)
    })
  })

  it('should cleanup on socket error', function (done) {
    const app = new Koa()
    var stream

    app.use(async (ctx) => {
      await send(ctx, '/test/fixtures/user.json')
      stream = ctx.body
      ctx.socket.emit('error', new Error('boom'))
    })

    request(app.listen())
      .get('/')
      .expect(500, function (err) {
        assert.ok(err)
        assert.ok(stream.destroyed)
        done()
      })
  })
})


test/experimental/async.js
'use strict';

var request = require('supertest');
var koa = require('koa');
var send = require('../../');

describe('asyncAwait support with koa experimental', function(){
    it('should 200', function(done){
        var app = koa();

        app.experimental = true;

        app.use(async function(){
            var ctx  = this;
            await send(ctx, 'test/fixtures/hello.txt');
        });

        request(app.listen())
        .get('/')
        .expect(200)
        .expect('world', done);
    });
});


src/send.utils.ts
/**
 * Module dependencies.
 */
import asyncFs from 'node:fs/promises';
import path from 'node:path';

/**
 * Returns `true` if the path exists, `false` otherwise.
 *
 * @param path
 * @return {boolean}
 * @api private
 */
export async function isPathExists(targetPath: string) {
  try {
    await asyncFs.access(targetPath);
    return true;
  } catch {
    return false;
  }
}

/**
 * Check if it's hidden.
 */
export function isPathHidden(root: string, targetPath: any) {
  const pathParts = targetPath.slice(root.length).split(path.sep);
  for (const part of pathParts) {
    if (part.at(0) === '.') return true;
  }

  return false;
}

/**
 * File type.
 */
export function getFileType(file: string, ext: string) {
  if (ext !== '') return path.extname(path.basename(file, ext));
  return path.extname(file);
}


index.js
/**
 * Module dependencies.
 */

const fs = require('fs')
const util = require('util')
const debug = require('debug')('koa-send')
const resolvePath = require('resolve-path')
const createError = require('http-errors')
const assert = require('assert')

const stat = util.promisify(fs.stat)
const access = util.promisify(fs.access)

async function exists (path) {
  try {
    await access(path)
    return true
  } catch (e) {
    return false
  }
}

const {
  normalize,
  basename,
  extname,
  resolve,
  parse,
  sep
} = require('path')

/**
 * Expose `send()`.
 */

module.exports = send

/**
 * Send file at `path` with the
 * given `options` to the koa `ctx`.
 *
 * @param {Context} ctx
 * @param {String} path
 * @param {Object} [opts]
 * @return {Promise}
 * @api public
 */

async function send (ctx, path, opts = {}) {
  assert(ctx, 'koa context required')
  assert(path, 'pathname required')

  // options
  debug('send "%s" %j', path, opts)
  const root = opts.root ? normalize(resolve(opts.root)) : ''
  const trailingSlash = path[path.length - 1] === '/'
  path = path.substr(parse(path).root.length)
  const index = opts.index
  const maxage = opts.maxage || opts.maxAge || 0
  const immutable = opts.immutable || false
  const hidden = opts.hidden || false
  const format = opts.format !== false
  const extensions = Array.isArray(opts.extensions) ? opts.extensions : false
  const brotli = opts.brotli !== false
  const gzip = opts.gzip !== false
  const setHeaders = opts.setHeaders

  if (setHeaders && typeof setHeaders !== 'function') {
    throw new TypeError('option setHeaders must be function')
  }

  // normalize path
  let decodedPath = decode(path)

  if (decodedPath === -1) return ctx.throw(400, 'failed to decode ' + path)

  // index file support
  if (index && trailingSlash) decodedPath += index

  decodedPath = resolvePath(root, decodedPath)

  // hidden file support, ignore
  if (!hidden && isHidden(root, decodedPath)) return

  let encodingExt = ''
  // serve brotli file when possible otherwise gzipped file when possible
  if (ctx.acceptsEncodings('br', 'identity') === 'br' && brotli && (await exists(decodedPath + '.br'))) {
    decodedPath = decodedPath + '.br'
    ctx.set('Content-Encoding', 'br')
    ctx.res.removeHeader('Content-Length')
    encodingExt = '.br'
  } else if (ctx.acceptsEncodings('gzip', 'identity') === 'gzip' && gzip && (await exists(decodedPath + '.gz'))) {
    decodedPath = decodedPath + '.gz'
    ctx.set('Content-Encoding', 'gzip')
    ctx.res.removeHeader('Content-Length')
    encodingExt = '.gz'
  }

  if (extensions && !/\./.exec(basename(decodedPath))) {
    const list = [].concat(extensions)
    for (let i = 0; i < list.length; i++) {
      let ext = list[i]
      if (typeof ext !== 'string') {
        throw new TypeError('option extensions must be array of strings or false')
      }
      if (!/^\./.exec(ext)) ext = `.${ext}`
      if (await exists(`${decodedPath}${ext}`)) {
        decodedPath = `${decodedPath}${ext}`
        break
      }
    }
  }

  // stat
  let stats
  try {
    stats = await stat(decodedPath)

    // Format the path to serve static file servers
    // and not require a trailing slash for directories,
    // so that you can do both `/directory` and `/directory/`
    if (stats.isDirectory()) {
      if (format && index) {
        decodedPath += `/${index}`
        stats = await stat(decodedPath)
      } else {
        return
      }
    }
  } catch (err) {
    const notfound = ['ENOENT', 'ENAMETOOLONG', 'ENOTDIR']
    if (notfound.includes(err.code)) {
      throw createError(404, err)
    }
    err.status = 500
    throw err
  }

  if (setHeaders) setHeaders(ctx.res, decodedPath, stats)

  // stream
  ctx.set('Content-Length', stats.size)
  if (!ctx.response.get('Last-Modified')) ctx.set('Last-Modified', stats.mtime.toUTCString())
  if (!ctx.response.get('Cache-Control')) {
    const directives = [`max-age=${(maxage / 1000 | 0)}`]
    if (immutable) {
      directives.push('immutable')
    }
    ctx.set('Cache-Control', directives.join(','))
  }
  if (!ctx.type) ctx.type = type(decodedPath, encodingExt)
  ctx.body = fs.createReadStream(decodedPath)

  return decodedPath
}

/**
 * Check if it's hidden.
 */

function isHidden (root, path) {
  path = path.substr(root.length).split(sep)
  for (let i = 0; i < path.length; i++) {
    if (path[i][0] === '.') return true
  }
  return false
}

/**
 * File type.
 */

function type (file, ext) {
  return ext !== '' ? extname(basename(file, ext)) : extname(file)
}

/**
 * Decode `path`.
 */

function decode (path) {
  try {
    return decodeURIComponent(path)
  } catch (err) {
    return -1
  }
}
